Add __int__, __float__, and __bool__ to the array object (#127)
data-apis/array-api: spec/API_specification/array_object.md

Element-wise results must equal the results returned by the equivalent element-wise function [`bitwise_and(x1, x2)`](elementwise_functions.md#logical_andx1-x2-).
```

(method-__bool__)=

> ### \_\_int\_\_(x, /)

Reviewer: ```suggestion
### \_\_bool\_\_(x, /)
```

> Converts a dimension-0 boolean array into a Python `bool` object.

Reviewer: ```suggestion
Converts a zero-dimensional boolean array to a Python `bool` object.
```

The general language across the spec has been to use "zero-dimensional" to refer to rank `0` arrays.

#### Parameters

-   **x**: _&lt;array&gt;_

>     -   dimension 0 array. Must have boolean data type.

Reviewer: ```suggestion
    -   zero-dimensional array. Must have boolean data type.
```

#### Returns

-   **out**: _&lt;bool&gt;_

>     -   a Python `bool` object representing the single element of the array `x`.

Reviewer 2: What should the behavior be if the array does not meet the requirement (no zero-dimensional, or different type)?.

Does the standard care whether `ValueError` or `TypeError` are raised?

NumPy can raise both:

```
In [2]: np.__version__
Out[2]: '1.20.0'

In [3]: X = np.zeros((2,2),dtype='d')

In [4]: X.__bool__
Out[4]: <method-wrapper '__bool__' of numpy.ndarray object at 0x2b67bdea50f0>

In [5]: X.__bool__()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-2acecd4ea032> in <module>
----> 1 X.__bool__()

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [6]: X.__float__()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-78817c2ae92b> in <module>
----> 1 X.__float__()

TypeError: only size-1 arrays can be converted to Python scalars

```

Reviewer 3: > Does the standard care whether `ValueError` or `TypeError` are raised?

In general we've refrained from specifying this, because (a) error handling is broad, (b) existing libraries are very inconsistent, and (c) there are situations where error handling strategy for accelerators may be different from that on CPU on purpose.

Element-wise results must equal the results returned by the equivalent element-wise function [`equal(x1, x2)`](elementwise_functions.md#equalx1-x2-).
```

(method-__float__)=

> ### \_\_int\_\_(x, /)

Reviewer: ```suggestion
### \_\_float\_\_(x, /)
```

> Converts a dimension-0 floating-point array into a Python `float` object.

Reviewer: ```suggestion
Converts a zero-dimensional floating-point array to a Python `float` object.
```

#### Parameters

-   **x**: _&lt;array&gt;_

>     -   dimension 0 array. Must have an floating-point data type.

Reviewer: ```suggestion
    -   dimension 0 array. Must have a floating-point data type.
```

Element-wise results must equal the results returned by the equivalent element-wise function [`greater(x1, x2)`](elementwise_functions.md#greaterx1-x2-).
```

(method-__int__)=
### \_\_int\_\_(x, /)

> Converts a dimension-0 integer array into a Python `int` object.

Reviewer: ```suggestion
Converts a zero-dimensional integer array to a Python `int` object.
```

#### Parameters

-   **x**: _&lt;array&gt;_

>     -   dimension 0 array. Must have an integer data type.

Reviewer: ```suggestion
    -   zero-dimensional array instance. Must have an integer data type.
```

#### Returns

-   **out**: _&lt;int&gt;_

    -   a Python `int` object representing the single element of the array `x`.

> 

Reviewer: ```suggestion
```
